### ISSUE TITLE
[BUG] _check_tree(has_aux=True) does not work

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#13092]
### TLDR

This line [in `_check_tree`](https://github.com/google/jax/blob/91d134d65b39c5570800fd788c3d41fac87ccebf/jax/_src/lax/control_flow/common.py#L117) is wrong, and should be replaced by 
```python
      actual_tree = actual_tree_children[0]
```

because those are already pytrees.

#### Longer explanation

I have an implementation of sparse solvers that return convergence information [here](https://gist.github.com/PhilipVinc/34e15d4fc93ee58753c24b29a8fba56e). Those work correctly if you pass in dense matrices and vectors in the sparse solver, but they give the following error if you try to use PyTrees as the vector:

```python
File ~/Documents/pythonenvs/netket/python-3.10.6/lib/python3.10/site-packages/jax/_src/lax/control_flow/common.py:125, in _check_tree(func_name, expected_name, actual_tree, expected_tree, has_aux)
    119     raise ValueError(
    120       f"{func_name}() produced a pytree with structure "
    121       f"{actual_tree}, but a pytree tuple with auxiliary "
    122       f"output was expected because has_aux was set to True.")
    124 if actual_tree != expected_tree:
--> 125   raise TypeError(
    126       f"{func_name}() output pytree structure must match {expected_name}, "
    127       f"got {actual_tree} and {expected_tree}.")

TypeError: solve() output pytree structure must match b, got PyTreeDef(*) and PyTreeDef(CustomNode(FrozenDict[()], [{'Dense': {'bias': *, 'kernel': *}, 'visible_bias': *}])).
```

But I am convinced that the objects I am passing in have the correct shape.

For example, if you run [this gist](https://gist.github.com/PhilipVinc/728a4a35f5d98713c384ab6f91dd6d3c) (which requires netket) where I'm using `cg` to solve the system of a fisher-information matrix and a set of parameters, this fails with the error above, even if it should not, because the code runs fine with jax's standard solvers.

I'm sorry for the long reproducer, but that was the most fast thing I could think of.

I tried doing the edit above and it works correctly so I think this is just a typo in jax itself.

### What jax/jaxlib version are you using?

master

**Comment by rajasekharporeddy** [comment]
Hi @PhilipVinc 

I can see that the related [PR #13093](https://github.com/google/jax/pull/13093) is merged. I tried to execute the provided code with latest JAX version 0.4.23 on Google Colab (CPU) and it executed without any error.

Kindly find the [gist](https://colab.research.google.com/gist/rajasekharporeddy/02442fe8752ad04820f582c51a876e8c/-13092.ipynb) for your reference.

Thank you.

**Comment by jakevdp** [comment]
Thanks for checking on this @rajasekharporeddy!
